Core/Src/main_task.c

```diff
@@ -234,27 +298,29 @@ static void shifting_task() {
 
 		// start a shift if there is one pending. This means that a new
 		// shift can be queued during the last shift
-		if(pending_shift == UPSHIFT) {
+		if(tcm_data.pending_shift == UPSHIFT) {
 			if (calc_validate_upshift(tcm_data.current_gear, tcm_data.sw_fast_clutch, tcm_data.sw_slow_clutch))
 			{
 				main_state = ST_HDL_UPSHIFT;
-				upshift_state = ST_U_BEGIN_SHIFT;
+				next_upshift_state = ST_U_BEGIN_SHIFT;
 			}
-		} else if (pending_shift == DOWNSHIFT) {
+		} else if (tcm_data.pending_shift == DOWNSHIFT) {
 			if (calc_validate_downshift(tcm_data.current_gear, tcm_data.sw_fast_clutch, tcm_data.sw_slow_clutch))
 			{
 				main_state = ST_HDL_DOWNSHIFT;
-				downshift_state = ST_D_BEGIN_SHIFT;
+				next_downshift_state = ST_D_BEGIN_SHIFT;
 			}
 		}
-		pending_shift = NONE;
+		tcm_data.pending_shift = NONE;
 		break;
 
 	case ST_HDL_UPSHIFT:
+		upshift_state = next_upshift_state; // Set state to next state variable per Sebastion's request
```

Comment:
Nit: You should set the state variable with the next value after you run the state machine so the first time you run you dont run the next_upshift_state